### PR TITLE
[FEATURE] Allow to configure location of UI5 home directory

### PR DESCRIPTION
**Pull request by d3xter666**
JIRA: CPOUI5FOUNDATION-703

### CODE REVIEW

**Review by flovogt — CHANGES_REQUESTED** [review]

File: lib/config/Configuration.js
Comment (d3xter666) [reply lib/config/Configuration.js:38]: Thanks!

File: test/lib/config/Configuration.js
Comment (d3xter666) [reply test/lib/config/Configuration.js:31]: My mistake. I was debugging something and forgot to restore it. Thanks!

File: lib/graph/helpers/ui5Framework.js
Comment (flovogt) [inline lib/graph/helpers/ui5Framework.js:376]: so now, we have the config `ui5-data-dir` but we map it directly to `ui5-home-dir`. Is this intended?

File: lib/config/Configuration.js
Comment (flovogt) [inline lib/config/Configuration.js:38]: what do we define as "artefacts"? these are the framework libs isn'´t?

Comment (d3xter666) [reply lib/config/Configuration.js:38]: By artifacts, I mean everything. It's basically the `.ui5` directory whith the following structure inside:
```yaml
- framework
   - artifacts
   - locks
   - metadata
   - packages
   - staging
```

Maybe, a better wording should be chosen? 🤔

Comment (flovogt) [reply lib/config/Configuration.js:38]: i would propose "framework artefacts"

File: lib/config/Configuration.js
Comment (flovogt) [inline lib/config/Configuration.js:21]: currently the configuration does not offer a default value but in real the home dir has a default value.... should we not align this?
so moving the defaulting also to config here?

File: lib/config/Configuration.js
Comment (d3xter666) [reply lib/config/Configuration.js:38]: Thanks!

File: lib/config/Configuration.js
Comment (d3xter666) [reply lib/config/Configuration.js:21]: Thanks for bringin it up!
This would mean that the `.ui5rc` file would always receive a value for the `ui5DataDir` and this setting would always be in there, regardless of the fact it's set intentionally or not.
Do we want to always have this in the `.ui5rc` file?

Also, the resolvers (i.e. `Openui5Resolver.js`) that utilize this directory are exposed through the `package.json` and in the best case we would need to double maintain the `ui5HomeDir`- in the `AbstractResolver.js` and in the `Configuration.js`. 
We cannot delete the default value from the `AbstractResolver.js (-> Openui5Resolver.js)` where it's populated currently, because that would be an incompatible change.

What do you think? Which way to choose? :D

Comment (flovogt) [reply lib/config/Configuration.js:21]: @RandomByte any opinion on the preferred way?

Comment (RandomByte) [reply lib/config/Configuration.js:21]: Good question. I think we discussed this as part of [RFC 0013 Configuration](https://github.com/SAP/ui5-tooling/blob/main/rfcs/0013-configuration.md#configuration-module)

There it says:
> [...] **Parameters shall not have default values.** [...]

So I think we should not add a defaulting here.

This means that Configuration.js is not the single source of truth for which ui5DataDir is used, but that's our compromise. Every consumer needs to apply the defaulting themselves, or we add a dedicated class/function somewhere to centralize that.

Actually, the environment variable must also not be read and applied in Configuration.js but on the consumer side. As per the same RFC:
> There shall **not be any fallback mechanism to environment variables** or similar inside the Configuration.js module. [...]

Comment (d3xter666) [reply lib/config/Configuration.js:21]: Ah, thanks a lot for clearing this out!

File: lib/config/Configuration.js
Comment (flovogt) [inline lib/config/Configuration.js:38]:
```suggestion
	 * Configurable directory where the framework artefacts are stored.
```

**Review by flovogt — CHANGES_REQUESTED** [review]

**Review by flovogt — COMMENTED** [review]
LGTM, waiting for review by @matz3 @RandomByte

File: lib/graph/helpers/ui5Framework.js
Comment (RandomByte) [inline lib/graph/helpers/ui5Framework.js:371]: I would expect that we use `UI5_DATA_DIR` as the "home dir" and don't create a `.ui5` directory inside of it.

**Review by RandomByte — CHANGES_REQUESTED** [review]

File: lib/graph/helpers/ui5Framework.js
Comment (d3xter666) [reply lib/graph/helpers/ui5Framework.js]: Done

**Review by RandomByte — APPROVED** [review]

**Pull request merged**
